src/nvim/api/window.c

```diff
@@ -33,6 +34,44 @@ Buffer nvim_win_get_buf(Window window, Error *err)
   return win->w_buffer->handle;
 }
 
+/// Sets the current buffer in a window. This allows switching the buffer of
+/// any window from within any other window without triggering side effects
+/// usually associated with switching between windows, such as triggering
+/// autocommands.
+///
+/// @param window   Window handle
+/// @param buffer   Buffer handle
+/// @param[out] err Error details, if any
+void nvim_win_set_buf(Window window, Buffer buffer, Error *err)
+  FUNC_API_SINCE(5)
+{
+  win_T *win = find_window_by_handle(window, err), *save_curwin = curwin;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+  tabpage_T *tab = win_find_tabpage(win), *save_curtab = curtab;
+
+  if (!win || !buf) {
+      return;
+  }
+
+  if (switch_win(&save_curwin, &save_curtab, win, tab, false) == FAIL) {
+    api_set_error(err,
+                  kErrorTypeException,
+                  "Failed to switch to window %d",
+                  window);
```

Comment:
should also test all of the failure modes (including the `!win || !buf` case above). In the past we've had hidden bugs until we added test cases for failure modes.

It's easy to test the error cases using `expect_err()`. See https://github.com/neovim/neovim/blob/8fd092f3ff15bf70f84ec0d716c5aaa2c7379fa1/test/functional/api/vim_spec.lua#L28                               
 for example.

Reply:
I have added tests for invalid window- and buffer ID, but as you said, I have no idea how to make the swap intentionally fail.